go/vt/throttler/demo/throttler_demo.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2016, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"flag"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/youtube/vitess/go/vt/discovery"
+	"github.com/youtube/vitess/go/vt/logutil"
+	"github.com/youtube/vitess/go/vt/tabletmanager/tmclient"
+	"github.com/youtube/vitess/go/vt/tabletserver/grpcqueryservice"
+	"github.com/youtube/vitess/go/vt/tabletserver/queryservice/fakes"
+	"github.com/youtube/vitess/go/vt/throttler"
+	"github.com/youtube/vitess/go/vt/vttest/fakesqldb"
+	"github.com/youtube/vitess/go/vt/wrangler"
+	"github.com/youtube/vitess/go/vt/wrangler/testlib"
+	"github.com/youtube/vitess/go/vt/zktopo/zktestserver"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+
+	log "github.com/golang/glog"
+)
+
+// This file contains a demo binary that demonstrates how the resharding
+// throttler adapts its throttling rate to the replication lag.
+//
+// The throttler is necessary because replicas apply transactions at a slower
+// rate than masters and fall behind at high write throughput.
+// (Mostly they fall behind because MySQL replication is single threaded but
+//  the write throughput on the master does not have to.)
+//
+// This demo simulates a client (writer), a master and a replica.
+// The client writes to the master which in turn replicas everything to the
+// replica.
+// The replica measures its replication lag via the timestamp which is part of
+// each message.
+// While the master has no rate limit, the replica is limited to
+// --rate (see below) transactions/second. The client runs the resharding
+// throttler which tries to throttle the client based on the observed
+// replication lag.
```

Comment:
Would be nice to be able to inject a few seconds delay at some point in there, see how we react to the spike.